File: python/msg.py
# Copyright (c) 2025 Ziqi Fan
# SPDX-License-Identifier: Apache-2.0

"""
ROS2 Message Type Definitions (using pycdr2)

Dependency: pip install pycdr2

Supports all ROS2 message types:
  - Basic types: bool, int8~64, uint8~64, float32, float64
  - Strings: str
  - Fixed arrays: array[T, N]
  - Dynamic arrays: sequence[T]
  - Nested structures: Direct composition

Examples:

# Basic types
@dataclass
class BasicTypes(IdlStruct):
    flag: bool
    count: int32
    value: float64

# String
@dataclass
class StringMsg(IdlStruct):
    name: str
    description: str

# Fixed array
@dataclass
class FixedArrayMsg(IdlStruct):
    data: array[float32, 10]
    xyz: array[float64, 3]

# Dynamic array
@dataclass
class DynamicArrayMsg(IdlStruct):
    values: sequence[int32]
    names: sequence[str]

# Nested structure
@dataclass
class Point(IdlStruct):
    x: float64
    y: float64
    z: float64

@dataclass
class Pose(IdlStruct):
    position: Point
    orientation: Point

# Complex message with all features
@dataclass
class ComplexMsg(IdlStruct):
    id: int32
    name: str
    matrix: array[array[float32, 3], 3]
    trajectory: sequence[Point]
"""

from dataclasses import dataclass

from pycdr2 import IdlStruct
from pycdr2.types import (  # noqa: F401
    array,
    float32,
    float64,
    int8,
    int16,
    int32,
    int64,
    sequence,
    uint8,
    uint16,
    uint32,
    uint64,
)


# geometry_msgs/msg/Vector3
@dataclass
class Vector3(IdlStruct):
    x: float64
    y: float64
    z: float64


# geometry_msgs/msg/Twist
@dataclass
class Twist(IdlStruct):
    linear: Vector3
    angular: Vector3

File: rust/src/msg.rs
// Copyright (c) 2025 Ziqi Fan
// SPDX-License-Identifier: Apache-2.0

//! ROS2 Message Type Definitions
//!
//! CDR serialization using serde + cdr crate
//!
//! Supports all ROS2 message types:
//!   - Basic types: bool, i8/u8, i16/u16, i32/u32, i64/u64, f32, f64
//!   - Strings: String
//!   - Fixed arrays: [T; N]
//!   - Dynamic arrays: Vec<T>
//!   - Nested structures: Direct composition
//!
//! Examples:
//!
//! // Basic types
//! #[derive(Serialize, Deserialize)]
//! struct BasicTypes {
//!     flag: bool,
//!     count: i32,
//!     value: f64,
//! }
//!
//! // String
//! #[derive(Serialize, Deserialize)]
//! struct StringMsg {
//!     name: String,
//!     description: String,
//! }
//!
//! // Fixed array
//! #[derive(Serialize, Deserialize)]
//! struct FixedArrayMsg {
//!     data: [f32; 10],
//!     xyz: [f64; 3],
//! }
//!
//! // Dynamic array
//! #[derive(Serialize, Deserialize)]
//! struct DynamicArrayMsg {
//!     values: Vec<i32>,
//!     names: Vec<String>,
//! }
//!
//! // Nested structure
//! #[derive(Serialize, Deserialize)]
//! struct Point { x: f64, y: f64, z: f64 }
//!
//! #[derive(Serialize, Deserialize)]
//! struct Pose {
//!     position: Point,
//!     orientation: Point,
//! }
//!
//! // Complex message with all features
//! #[derive(Serialize, Deserialize)]
//! struct ComplexMsg {
//!     id: i32,
//!     name: String,
//!     matrix: [[f32; 3]; 3],
//!     trajectory: Vec<Point>,
//! }

#![allow(dead_code)]

use serde::{Deserialize, Serialize};

// geometry_msgs/msg/Vector3
#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Vector3 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

// geometry_msgs/msg/Twist
#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Twist {
    pub linear: Vector3,
    pub angular: Vector3,
}
